### ISSUE TITLE
The order of locking in azurerm_network_interface could lead to perpetual lock in some scenario

### ISSUE THREAD

**Issue by hashitop** [terraform-providers/terraform-provider-azurerm#13143]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Affected Resource(s)

* `azurerm_network_interface`

### Terraform Configuration Files

The configuration file is no longer available but the relevant source code from the resource is located here:

https://github.com/hashicorp/terraform-provider-azurerm/blob/v2.73.0/internal/services/network/network_interface_locking.go#L15-L18

The resource should attempt to acquire the lock when the `ip_configuration` is specified.

https://github.com/hashicorp/terraform-provider-azurerm/blob/v2.73.0/internal/services/network/network_interface_resource.go#L237-L247

In comparison to the order of lock operation in `azurerm_subnet_network_security_group_association` resource from the recent change in PR #12267, the order of the lock operation can lead to the perpetual lock at random times.

### Panic Output

N/A

### Expected Behaviour

Apply operation completed successfully

### Actual Behaviour

Apply operation got stuck when these resources started:

```
module.xxx_prod_vm_subnet.azurerm_subnet_network_security_group_association.association
module.xxx_prod_vm_subnet.azurerm_subnet_route_table_association.spoke-prod-ase[0]
module.xxx_pt_vm_subnet.azurerm_subnet_network_security_group_association.association
module.xxx_pt_vm_subnet.azurerm_subnet_route_table_association.spoke-prod-ase[0]
module.xxx-prod-sql-instance.module.subnet_primary.azurerm_subnet_network_security_group_association.association
module.xxx-prod-sql-instance.module.subnet_primary.azurerm_subnet_route_table_association.udr[0]
module.xxx-prod-windows-vm-1.azurerm_network_interface.main[0]
module.xxx-pt-sql-instance.module.subnet_primary.azurerm_subnet_network_security_group_association.association
module.xxx-pt-sql-instance.module.subnet_primary.azurerm_subnet_route_table_association.udr[0]
module.xxx-pt-windows-vm-2.azurerm_network_interface.main[0]
```

### Steps to Reproduce

The scenario is difficult to reproduce

**Comment by rodrigotmedeiros** [comment]
Any update on this thread? 

I'm currently having this issue: 
azurerm_network_interface.myvmnic["ANSIBLE-UBUNTU"]: Still creating... [4m50s elapsed]
azurerm_subnet_network_security_group_association.associacao_nsg: Still creating... [4m50s elapsed]

**Comment by yeoldegrove** [comment]
I had the very same issue starting with `2.82.0`. At first glance, it seems to be fixed in `2.97.0` for me.

**Comment by rcskosir** [comment]
Thanks for opening this issue. This was a problem in the 2.x version of the provider which is no longer actively maintained.
If this is still an issue with the 3.x version of the provider please do let us know by opening a new issue, thanks!
